paddle/operators/detail/grpc_server.h

```diff
@@ -0,0 +1,97 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include "paddle/framework/lod_tensor.h"
+#include "paddle/framework/scope.h"
+#include "paddle/framework/selected_rows.h"
+#include "paddle/framework/var_type.h"
+#include "paddle/operators/detail/simple_block_queue.h"
+
+#include "paddle/operators/detail/send_recv.grpc.pb.h"
+#include "paddle/operators/detail/send_recv.pb.h"
+
+#include <grpc++/grpc++.h>
+#include <grpc/support/log.h>
+#include <thread>
+#include "paddle/operators/detail/utils.h"
+
+namespace paddle {
+namespace operators {
+namespace detail {
+
+typedef std::pair<std::string, sendrecv::VariableMessage> MessageWithName;
+class RequestBase;
+
+class AsyncGRPCServer final : public sendrecv::SendRecvService::Service {
+ public:
+  explicit AsyncGRPCServer(std::string address) { address_ = address; }
+
+  void RunSyncUpdate();
+
+  void Reset();
+
+  void Done();
+
+  void SetScope(framework::Scope *scope) { scope_ = scope; }
+
+  const MessageWithName Get() { return this->var_recv_queue_.Pop(); }
+
+  void Push(const MessageWithName &msg) { this->var_recv_queue_.Push(msg); }
+
+  void ShutDown();
+
+ protected:
+  void Wait();
+  void HandleReqSend();
+  void HandleReqGet(bool wait);
+  // void Proceed(void *tag, bool ok);
+  void TryToRegisterNewGet();
+  void TryToRegisterNewSend();
+  void SetGetFinishOrDelete(RequestBase *&last);
+  void SetSendFinishOrDelete(RequestBase *&last);
+
+  void ShutdownGetQueue();
+  void ShutdownSendQueue();
+
+ private:
+  std::mutex cq_send_mutex_;
+  mutable bool is_send_shut_down_ = false;
+  std::unique_ptr<grpc::ServerCompletionQueue> cq_send_;
+
+  std::mutex cq_get_mutex_;
+  mutable bool is_get_shut_down_ = false;
+  std::unique_ptr<grpc::ServerCompletionQueue> cq_get_;
```

Comment:
Can server-side use one single `ServerCompletionQueue`? And two threads to deal with different message make this implement complicated. Since we have `var_recv_queue_` so the backend can use multiple threads to deal with incoming requests.

Reply:
Resolved in another way.Done.